internal/testutils/broker.go

```diff
@@ -0,0 +1,278 @@
+
+type isAuthorizedCtx struct {
+	ctx        context.Context
+	cancelFunc context.CancelFunc
+}
+
+// BrokerBusMock is the D-Bus object that will answer calls for the broker mock.
+type BrokerBusMock struct {
+	name                string
+	isAuthorizedCalls   map[string]isAuthorizedCtx
+	isAuthorizedCallsMu sync.Mutex
+}
+
+// StartBusBrokerMock starts the D-Bus service and exports it on the system bus.
+// It returns the configuration file path for the exported broker.
+func StartBusBrokerMock(t *testing.T) string {
+	t.Helper()
+
+	brokerName := strings.ReplaceAll(t.Name(), "/", "_")
+	busObjectPath := fmt.Sprintf(objectPathFmt, brokerName)
+	busInterface := fmt.Sprintf(interfaceFmt, brokerName)
+
+	conn, err := dbus.ConnectSystemBus()
+	require.NoError(t, err, "Setup: could not connect to system bus")
+	t.Cleanup(func() { require.NoError(t, conn.Close(), "Teardown: could not close system bus connection") })
+
+	bus := BrokerBusMock{
+		name:                brokerName,
+		isAuthorizedCalls:   map[string]isAuthorizedCtx{},
+		isAuthorizedCallsMu: sync.Mutex{},
+	}
+	err = conn.Export(&bus, dbus.ObjectPath(busObjectPath), busInterface)
+	require.NoError(t, err, "Setup: could not export mock broker")
+
+	err = conn.Export(introspect.NewIntrospectable(&introspect.Node{
+		Name: busObjectPath,
+		Interfaces: []introspect.Interface{
+			introspect.IntrospectData,
+			{
+				Name:    busInterface,
+				Methods: introspect.Methods(&bus),
+			},
+		},
+	}), dbus.ObjectPath(busObjectPath), introspect.IntrospectData.Name)
+	require.NoError(t, err, "Setup: could not export mock broker introspection")
+
+	reply, err := conn.RequestName(busInterface, dbus.NameFlagDoNotQueue)
+	require.NoError(t, err, "Setup: could not request mock broker name")
+	require.Equal(t, reply, dbus.RequestNameReplyPrimaryOwner, "Setup: mock broker name already taken")
+
+	configPath := writeConfig(t, brokerName)
+
+	t.Cleanup(func() {
+		r, err := conn.ReleaseName(busInterface)
+		require.NoError(t, err, "Teardown: could not release mock broker name")
+		require.Equal(t, r, dbus.ReleaseNameReplyReleased, "Teardown: mock broker name not released")
+	})
+
+	return configPath
+}
+
+func writeConfig(t *testing.T, name string) string {
+	t.Helper()
+
+	cfgPath := filepath.Join(t.TempDir(), "broker-cfg")
+
+	s := fmt.Sprintf(brokerConfigTemplate, name, name, name, name)
+	err := os.WriteFile(cfgPath, []byte(s), 0600)
+	require.NoError(t, err, "Setup: Failed to write broker config file")
+
+	return cfgPath
+}
+
+// NewSession returns default values to be used in tests or an error if requested.
+func (b *BrokerBusMock) NewSession(username, lang string) (sessionID, encryptionKey string, dbusErr *dbus.Error) {
+	if username == "NS_error" {
+		return "", "", dbus.MakeFailedError(fmt.Errorf("Broker %q: NewSession errored out", b.name))
+	}
+	if username == "NS_no_id" {
+		return "", username + "_key", nil
+	}
+	return fmt.Sprintf("%s-%s_session_id", b.name, username), username + "_key", nil
```

Comment:
I would change the encryption key to be only based on the broker (actually, I expected it to only have one per broker)